File: SSUMobile/Modules/Core/SSUConfiguration.m
//
//  SSUConfiguration.m
//  SSUMobile
//
//  Created by Eric Amorde on 9/16/15.
//  Copyright (c) 2015 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUConfiguration.h"
#import "SSUCommunicator.h"
#import "SSULogging.h"

static NSString * const kSSUConfigKeyPrefix = @"edu.sonoma";
NSString * const kSSUConfigLastLoadDateKey = @"edu.sonoma.configuration.last_load";

@interface SSUConfiguration()

@property (nonatomic, strong) NSUserDefaults * userDefaults;

@end

@implementation SSUConfiguration

+ (instancetype) sharedInstance {
    static SSUConfiguration * instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

- (NSUserDefaults *) userDefaults {
    if (_userDefaults) return _userDefaults;
    
    _userDefaults = [NSUserDefaults standardUserDefaults];
    
    return _userDefaults;
}

#pragma mark - Helper

- (void) loadFromURL:(NSURL *)url completion:(void (^)(NSError *))completion {
    NSDate * lastLoadDate = [NSDate date];
    [SSUCommunicator getJSONFromURL:url completion:^(NSURLResponse *response, id json, NSError *error) {
        if (error) {
            SSULogError(@"Error while attempting to load settings from remote url: %@", error);
            if (completion)
                completion(error);
        }
        else if (![json isKindOfClass:[NSDictionary class]]) {
            SSULogError(@"Expected dictionary from JSON, got %@ instead", NSStringFromClass([json class]));
            if (completion)
                completion(nil);
        }
        else {
            [self loadDictionary:json];
            [self setDate:lastLoadDate forKey:kSSUConfigLastLoadDateKey];
            if (completion)
                completion(nil);
        }
    }];
}

- (void) loadDictionary:(NSDictionary *)data {
    [data enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
        NSAssert([key isKindOfClass:[NSString class]], @"All config keys should be strings");
        if (obj == [NSNull null]) {
            [self removeObjectForKey:key];
        }
        else {
            [self setObject:obj forKey:key];
        }
    }];
}

- (void) loadDefaultsFromFilePath:(NSString *)filePath {
    NSData * defaultsData = [NSData dataWithContentsOfFile:filePath];
    NSError * error;
    NSDictionary * defaults = [NSJSONSerialization JSONObjectWithData:defaultsData options:0 error:&error];
    if (error) {
        SSULogError(@"Error while loading JSON for defaults: %@", error);
    }
    else {
        [self registerDefaults:defaults];
    }
}

- (void) registerDefaults:(NSDictionary *)defaults {
    [self.userDefaults registerDefaults:defaults];
}

- (void) save {
    [self.userDefaults synchronize];
}

#pragma mark - Accessors

- (id) objectForKey:(NSString *)key {
    return [self.userDefaults objectForKey:key];
}

- (NSString *) stringForKey:(NSString *)key {
    return [self.userDefaults stringForKey:key];
}

- (NSDate *) dateForKey:(NSString *)key {
    return [self.userDefaults objectForKey:key];
}

- (NSArray *) arrayForKey:(NSString *)key {
    return [self.userDefaults arrayForKey:key];
}

- (NSDictionary *) dictionaryForKey:(NSString *)key {
    return [self.userDefaults dictionaryForKey:key];
}

- (NSData *) dataForKey:(NSString *)key  {
    return [self.userDefaults dataForKey:key];
}

- (NSArray <NSString *> *) stringArrayForKey:(NSString *)key {
    return [self.userDefaults stringArrayForKey:key];
}

- (NSInteger) integerForKey:(NSString *)key {
    return [self.userDefaults integerForKey:key];
}

- (float) floatForKey:(NSString *)key {
    return [self.userDefaults floatForKey:key];
}

- (double) doubleForKey:(NSString *)key {
    return [self.userDefaults doubleForKey:key];
}

- (BOOL) boolForKey:(NSString *)key {
    return [self.userDefaults boolForKey:key];
}

- (NSURL *) URLForKey:(NSString *)key {
    return [self.userDefaults URLForKey:key];
}

- (NSDictionary *) dictionaryRepresentation {
    NSDictionary * fullDictionary = [self.userDefaults dictionaryRepresentation];
    NSMutableDictionary * result = [NSMutableDictionary new];
    [fullDictionary enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
        if ([key rangeOfString:kSSUConfigKeyPrefix].location != NSNotFound) {
            result[key] = obj;
        }
    }];
    return result;
}

#pragma mark - Setters

- (void) setObject:(id)value forKey:(NSString *)key {
    [self.userDefaults setObject:value forKey:key];
}

- (void) setArray:(NSArray *)array forKey:(NSString *)key {
    [self.userDefaults setObject:array forKey:key];
}

- (void) setDate:(NSDate *)date forKey:(NSString *)key {
    [self.userDefaults setObject:date forKey:key];
}

- (void) setStringArray:(NSArray<NSString *> *)stringArray forKey:(NSString *)key {
    [self.userDefaults setObject:stringArray forKey:key];
}

- (void) setString:(NSString *)string forKey:(NSString *)key {
    [self setObject:string forKey:key];
}

- (void) setInteger:(NSInteger)value forKey:(NSString *)key {
    [self.userDefaults setInteger:value forKey:key];
}

- (void) setFloat:(float)value forKey:(NSString *)key {
    [self.userDefaults setFloat:value forKey:key];
}

- (void) setDouble:(double)value forKey:(NSString *)key {
    [self.userDefaults setDouble:value forKey:key];
}

- (void) setBool:(BOOL)value forKey:(NSString *)key {
    [self.userDefaults setBool:value forKey:key];
}

- (void) setURL:(NSURL *)value forKey:(NSString *)key {
    [self.userDefaults setURL:value forKey:key];
}

- (void)removeObjectForKey:(NSString *)key {
    [self.userDefaults removeObjectForKey:key];
}

@end

File: SSUMobile/Modules/Core/SSUConfiguration.swift
//
//  SSUConfiguration.swift
//  SSUMobile
//
//  Created by Eric Amorde on 6/4/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation

/**
 Stores basic configuration data on disk, persisting through multiple app launches.
 
 Supports any PLIST (String, Date, Array, Dictionary, Data) as well as some additional types such as URL.
 
 If the type of object you want to store is not supported, you can convert it to a Data object before storing it and then
 convert it from data when accessing it.
 
 The storage is backed by UserDefaults
 
 You can observe changes to individual keys with Key Value Observing
 */
@objc
class SSUConfiguration: NSObject {
    
    enum ConfigurationError: Error {
        case fileNotFound
    }
    
    private let userDefaults: UserDefaults
    
    @objc(sharedInstance)
    static let instance = SSUConfiguration(userDefaults: UserDefaults.standard)
    
    init(userDefaults: UserDefaults) {
        self.userDefaults = userDefaults
    }
    
    var dictionaryRepresentation: [String:Any] {
        return userDefaults.dictionaryRepresentation()
    }
    
    private let _lastLoadDateKey = "edu.sonoma.configuration.last_load"
    
    /**
     The date of the most recent loading of configuration data, either from a file or from a remote URL.
     
     This date does not reflect changes to individual keys
     */
    var lastLoadDate: Date? {
        set {
            set(newValue, forKey: _lastLoadDateKey)
        } get {
            return date(forKey: _lastLoadDateKey)
        }
    }
    
    private func withKVOChange(_ key: String, block: () -> Void) {
        willChangeValue(forKey: key)
        block()
        didChangeValue(forKey: key)
    }
    
    // MARK: Accessors
    
    func object(forKey key: String) -> Any? {
        return userDefaults.object(forKey: key)
    }
    
    func string(forKey key: String) -> String? {
        return userDefaults.string(forKey: key)
    }

    func date(forKey key: String) -> Date? {
        return userDefaults.object(forKey: key) as? Date
    }
    
    func data(forKey key: String) -> Data? {
        return userDefaults.data(forKey: key)
    }

    func stringArray(forKey key: String) -> [String]? {
        return userDefaults.stringArray(forKey: key)
    }

    func integer(forKey key: String) -> Int {
        return userDefaults.integer(forKey: key)
    }

    func float(forKey key: String) -> Float {
        return userDefaults.float(forKey: key)
    }

    func double(forKey key: String) -> Double {
        return userDefaults.double(forKey: key)
    }

    func bool(forKey key: String) -> Bool {
        return userDefaults.bool(forKey: key)
    }

    func url(forKey key: String) -> URL? {
        return userDefaults.url(forKey: key)
    }
    
    // MARK: Setters
    
    func set(_ object: Any?, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(object, forKey: key)
        }
    }
    
    @objc(setString:forKey:)
    func set(_ string: String?, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(string, forKey: key)
        }
    }

    @objc(setDate:forKey:)
    func set(_ date: Date?, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(object, forKey: key)
        }
    }
    
    @objc(setData:forKey:)
    func set(_ data: Data?, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(data as Any?, forKey: key)
        }
    }

    @objc(setStringArray:forKey:)
    func set(_ stringArray: [String]?, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(stringArray, forKey: key)
        }
    }

    @objc(setInteger:forKey:)
    func set(_ integer: Int, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(integer, forKey: key)
        }
    }

    @objc(setFloat:forKey:)
    func set(_ float: Float, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(float, forKey: key)
        }
    }
    
    @objc(setDouble:forKey:)
    func set(_ double: Double, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(double, forKey: key)
        }
    }

    @objc(setBool:forKey:)
    func set(_ bool: Bool, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(bool, forKey: key)
        }
    }

    @objc(setURL:forKey:)
    func set(_ url: URL?, forKey key: String) {
        withKVOChange(key) {
            userDefaults.set(url, forKey: key)
        }
        
    }
    
    /**
     Deletes any existing value for the given key. Functionally the same as setting it to nil
     */
    func removeObject(forKey key: String) {
        withKVOChange(key) { 
            userDefaults.removeObject(forKey: key)
        }
    }
    
    // MARK: Helper
    
    func registerDefaults(_ defaults: [String:Any]) {
        userDefaults.register(defaults: defaults)
    }
    
    func registerDefaults(fromFilename filename: String, completion: (() -> Void)? = nil) {
        guard let url = Bundle.main.url(forResource: filename, withExtension: nil) else {
            SSULogging.logError("Unable to retrieve resource url for filename: \(filename)")
            completion?()
            return
        }
        // Don't load this asynchronously because we need it immediately for setting up modules
        do {
            let data = try Data(contentsOf: url)
            let dict = try JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
            self.registerDefaults(dict)
        } catch {
            SSULogging.logError("Error while attempting to load defaults: \(error)")
            assert(false, "Unable to load defaults")
        }
    }

    func load(dict: [String:Any?]) {
        for (key, value) in dict {
            if value is NSNull {
                removeObject(forKey: key)
            } else {
                set(value, forKey: key)
            }
        }
        lastLoadDate = Date()
    }
    
    
    
    /**
     Loads configuration values from a dictionary provided by the given url.
     
     If the URL is a local file url, the file data will be loaded in the background and parsed as a dictionary.
     If the URL is an HTTP(S) URL, a network request will be performed and the response will be interpreted as JSON.
     */
    func loadFrom(url: URL, completion: ((Error?) -> Void)? = nil) {
        SSUCommunicator.getJSONFrom(url) { (response, obj, error) in
            defer {
                completion?(error)
            }
            if let error = error {
                SSULogging.logError("Error during config JSON loading: \(error)")
            } else if let dict = obj as? [String:Any] {
                self.load(dict: dict)
            } else {
                SSULogging.logError("Received unexpected type when loading configuration: \(type(of: obj))")
            }
        }
    }
    
    /**
     Loads from the given filename in the main bundle
     */
    func loadFrom(filename: String, completion: ((Error?) -> Void)? = nil) {
        guard let url = Bundle.main.url(forResource: filename, withExtension: nil) else {
            SSULogging.logError("Unable to find file \(filename)")
            completion?(ConfigurationError.fileNotFound)
            return
        }
        
        loadFrom(url: url, completion: completion)
    }
    
    /**
     Saves current values to disk.
     
     You normally do not need to call this method directly, as configuration is automatically saved periodically.
     - returns: `true` if successfully saved to disc, `false` otherwise
     */
    @discardableResult
    func save() -> Bool {
        return userDefaults.synchronize()
    }
}


/// Deprecated
extension SSUConfiguration {
    @nonobjc
    @available(*, deprecated, renamed: "instance")
    static func sharedInstance() -> SSUConfiguration {
        return instance
    }
    @nonobjc
    @available(*, unavailable, renamed: "set(_:forKey:)")
    func setDate(_ date: Date?, forKey key: String) {}
    @nonobjc
    @available(*, unavailable, renamed: "register(defaults:)")
    func registerDefaults(_ defaults: [String:Any?]) {}
}

